python/mxnet/module/base_module.py

```diff
@@ -333,9 +334,13 @@ def fit(self, train_data, eval_data=None, eval_metric='acc',
 
         if hasattr(train_data, 'layout_mapper'):
             self.layout_mapper = train_data.layout_mapper
-        data_types = dict(train_data.provide_data_type + train_data.provide_label_type)
-        self.bind(data_shapes=train_data.provide_data, label_shapes=train_data.provide_label,
-                  for_training=True, force_rebind=force_rebind, input_types=data_types)
+
+        provide_data = DataDesc.get_list(train_data.provide_data,
```

Comment:
do this in executor_manager so you don't need to do it in every module. Module's data_shapes should return what ever user passes to it to minimize confusion